### DISCUSSION TITLE
Theia-Tree-Editor adaptation to use any model.

### DISCUSSION THREAD

**Discussion by Reeze-Moodaley** [eclipse-emfcloud/emfcloud#123] [Theia Tree Editor] [answered]
Hi there we are currently working on a project where we would like to use the theia-tree-editor to display an XML model that we have on our Model Server, how would one go about adjusting the tree editor in a manner that would allow us to make use of this model?

**Comment by ndoschek** [comment, marked as answer]
Hi @Reeze-Moodaley,

We use the `theia-tree-editor` in combination with the Model Server in our [`coffee-editor` project](https://github.com/eclipse-emfcloud/coffee-editor).

There we use the `theia-tree-editor` combined with a JSONForms based editor to edit a subset of the [Coffee model](https://github.com/eclipse-emfcloud/coffee-editor/blob/master/backend/plugins/org.eclipse.emfcloud.coffee.model/model/Coffee.ecore).
For details please check out the [`coffee-editor-extension`](https://github.com/eclipse-emfcloud/coffee-editor/tree/master/web/coffee-editor-extension/src/browser/coffee-tree).

Here is a small sneak peek:
![coffee-tree-editor](https://user-images.githubusercontent.com/47414588/156398452-fa3e26f1-7a0c-4357-a99b-dbe9c5013311.gif)

Hope that helps, BR Nina

**Reply by Reeze-Moodaley** [reply]
Thank you! I'll have a look at it and get back to you.

**Reply by eneufeld** [reply]
@Reeze-Moodaley is the question answered for you?

**Reply by Reeze-Moodaley** [reply]
Hi there! 

Yes we managed to somewhat get the feature working.

**Comment by rsoika** [comment]
I have a similar question: I also want to use the JSON Forms. But not the TreeView. I would like to show the form when the user selects a model element in the modelling pane. I guess this is possible. 
I there any documentation how to integrate JSON forms on the client side? The videos are always so cool. But it's hell to find out how something works only through time intensive reverse engineering.

**Reply by ndoschek** [reply]
Hi @rsoika,
If you do not want to use the tree view, you could take a look at the [JSONForms property-view](https://github.com/eclipse-emfcloud/jsonforms-property-view). It also offers a Model Server aware version to fetch element data directly from there.
It reacts to GLSP selections and provides the (editable) fetched data in a JSON Forms widget.
You can find an example in our [ecore-glsp editor project](https://github.com/eclipse-emfcloud/ecore-glsp/tree/master/client/packages).

**Reply by rsoika** [reply]
@ndoschek  can you please provide a more general short guideline how to adapt the JSONForms property-view? 
I really try to understand the ecore-glsp project with the property view extension. But it consists of several hundred files with no comments at all.  Also it looks as if the project is entangled with other emf-cloud projects. So it's not always clear which parts a from the concrete example (ecore) and which are general concepts from GLSP or EMF-cloud
It makes it nearly impossible for me to extract the important parts and learn what the concepts are behind all this. For example:

- what is a `ModelserverAwareWidgetProvider` good for and how does it work? 
- the file `property-data-service.ts` seems to be very specific to the generic concepts behind ecore?   Do I need this file?
- I guess that the `PropertyViewWidgetProvider` is the important part. But how does it work in combination with GLSP?

I really appreciate your project. But I still don't understand why not a single comment is added in the source files. THis would help so  much and would be so little additional effort for the developer. I always start enthusiastic to try out new feature, and after some hours I'm just frustrated because I have no chance to understand just a bit.

Maybe we can create together a short doc how to create a super-simple JSONForms  property-view which just shows the element id and the name of a selected element from the diagram pane. I have already written such docs for [Ports](https://github.com/imixs/open-bpmn/blob/master/doc/PORTS.md) and the [ToolPalette](https://github.com/imixs/open-bpmn/blob/master/doc/TOOL_PALETTE.md)

**Reply by ndoschek** [reply]
Hi @rsoika,
I will move that thread into a separate discussion(https://github.com/eclipse-emfcloud/emfcloud/discussions/126) if you don't mind, as it is not directly linked to the `theia-tree-editor`.
